posts/2017-05-02-io-monad-for-cats.md

```diff
@@ -0,0 +1,404 @@
+
+The answer is that we want to reason about *where* and *when* our effects are evaluated.  And of course, we want to be able to leverage laws and abstractions which assume equational semantics for expressions (i.e. referential transparency).  Cats is full of these sorts of abstractions, and cats-laws provides a vast set of laws which describe them.  But all of these abstractions and all of these laws break down the *moment* you introduce some sort of side-effecting expression.  Because, much like our referential transparency example from earlier, these abstractions *assume* that you can substitute expressions with their evaluated results, and that's just not true in the presence of side-effects.
+
+What we need is a data type which allows us to encapsulate Scala-style side-effects in the form of a *pure* value, on which referential transparency holds and which we can compose using other well-defined abstractions, such as `Monad`.  Scalaz defines two such data types which meet these criteria: `scalaz.effect.IO` and `scalaz.concurrent.Task`.  But in practice, nearly everyone uses `Task` instead of `IO` because of its support for *asynchronous* effects.
+
+Cats does not define any such abstraction, and what's worse is the cats *ecosystem* also doesn't really provide any such abstraction.  There are two `Task` implementations that are relatively commonly used with cats – namely, `monix.Task` and `fs2.Task` – but these are not part of cats per se, nor are they deeply integrated into its abstraction hierarchy.  Additionally, the proliferation of broadly equivalent options has led to confusion in the ecosystem, as well as some justifiable fear, uncertainty and doubt on the part of those evaluating cats as a potential dependency.
+
+## Introducing cats-effect
+
+The [cats-effect](https://github.com/typelevel/cats-effect) project aims to change all of that.  The goal of cats-effect is to provide an "easy default" `IO` type for the cats ecosystem, deeply integrated with cats-core, with all of the features and performance that are required for real world production use.  Additionally, cats-effect defines a set of abstractions in the form of several typeclasses which describe what it means to *be* a pure effect type.  These abstractions are extremely useful both in enabling MTL-style program composition and to ensure that other pre-existing `Task` implementations remain first-class citizens of the ecosystem.  `IO` does not overshadow `monix.Task` or `fs2.Task`; it *complements* them by providing a set of abstractions and laws which allow users to write safe, parametric code which supports each of them equally.
```

Comment:
`monix.Task` -> `monix.eval.Task` 

(if the intention is to give the full path, like in `scalaz.concurrent.Task`, instead of `scalaz.Task`)